### ISSUE TITLE
Http4sRoutes Types Issue

### ISSUE THREAD

**Issue by SarpongAbasimi** [http4s/http4s#3952]
Please this works perfectly
even though the signature for `of` is of type ` of[F[_]: Defer: Applicative]` which means that an `IO` is of type `Defer` or `Applicative` ?
<img width="360" alt="Screen Shot 2020-11-27 at 21 14 41" src="https://user-images.githubusercontent.com/37377831/100486132-9b46d580-30fa-11eb-8dcc-3a65b484569c.png">
also this does not work and my answer is why even though F is of type `IO`

<img width="538" alt="Screen Shot 2020-11-27 at 21 21 08" src="https://user-images.githubusercontent.com/37377831/100486148-b31e5980-30fa-11eb-866d-2943ccaebea0.png">

Please have some patience with me if you are a seasoned Scala Dev. I just would like to understand why it works

**Comment by Daenyth** [comment]
tldr version: `def index: HttpRoutes[IO] = HttpRoutes.of[IO] { ...`

-----

The issue here is a misunderstanding of what `: Foo` means in a type argument. It's not an "is a" relationship. It's a usage of `implicit` parameters for a "has capability of" relationship.

`def index[F[_]: Defer: Applicative]` means "for some type constructor `F[_]` which has capabilities of `Defer` and `Applicative`.

Mechanically, this is using ["context bound"](https://docs.scala-lang.org/tutorials/FAQ/context-bounds.html) syntax sugar.

`def index[F[_]: Defer: Applicative]` is syntax sugar for `def index[F[_]](implicit ev1: Defer[F], ev2: Applicative[F])`.

-----

The reason your second example there with `F[_]: IO` doesn't work is because it becomes syntax sugar for `implicit ev1: IO[F]`, which is both incorrect conceptually, and also wrongly-kinded. "Kind" here as in "higher-kinded types". One article covering some of that concept is [this typelevel blog post](https://typelevel.org/blog/2016/08/21/hkts-moving-forward.html)

By "kind" here, I mean the "shape" of the type. `def foo[A]` means "for some type `A`". Examples of types with that shape (that "kind") are types with values, like `Int` and `String`.

Whereas `def foo[F[_]]` means "for some type that itself has one type argument". Often those are called "type constructors". Examples of things with that shape are `List` and `IO`. Note, not `List[Int]` or `IO[String]`, but *just* `List` and `IO`. One way to think about types of that shape is as type-level functions. Where a normal function takes a value input and returns a value output, a type constructor takes a *type* input and gives you a *type* output.

For example, the type constructor `List` has no values. Only kinds with no arguments have values (like `Int`, or more generally, `[A]`-shaped things). You give `List` the type argument of `Int`, and you get the *type* `List[Int]`, which can have values like `List(1, 2)`.

-----

Other terms that you'll come across in this space are "tagless final" design, and the ["typeclass" pattern](https://books.underscore.io/essential-scala/essential-scala.html#sec:type-classes).

Both of those techniques are ways to *very generally* apply the "code to an interface, and not an implementation" philosophy of java. Instead of working with *only* `IO`, you could work with other types as well, like `monix.Task` or `EitherT[IO, MyAppErrorTypes, *]`. I won't go into more detail here, but I will say that "tagless final" design is a more advanced technique that I recommend you avoid until you're more comfortable with it. It requires a higher learning curve both to read and write, and requires that people working in that code learn some concepts which are almost certainly new to them, since most languages are unable to express this kind of design pattern.


Thanks for the question, and please feel free to ask more :)
Folks are also often on the gitter chat - we're always happy to have newcomers to the community if you want to hang out or have some discussion / ask for assistance.

**Comment by SarpongAbasimi** [comment]
Thank you very much @Daenyth for your time, this is the best explanation I have ever had.
Everything makes sense now.
